File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章日志统计.cpp
//日志统计 2018年第九届省赛C/C++大学A组  lanqiaoOJ 179 

#include<bits/stdc++.h>
using namespace std;
const int N = 1e5+50;
int num[N];    //num[i]：记录id=i的帖子的赞的数量
int flag[N];   //flag[i]：id=i的贴子曾是热帖
struct post{int id, ts;}p[N];       //记录帖子
int cmp(post x,post y){ return x.ts < y.ts;  } //按时间从小到大排序
int main(){
    int n,d,k; cin>>n>>d>>k;
    for(int i=0;i<n;i++) scanf("%d%d",&p[i].ts,&p[i].id);
    sort(p,p+n,cmp);            //按时间从小到大排序
    for(int i=0,j=0;i<n;i++){
        num[p[i].id]++;
        while(p[i].ts - p[j].ts >= d){
            num[p[j].id]--;     //随着时间流逝，d之前的每个贴的次数都减1
            j++;
        }
        if(num[p[i].id] >= k)   //在区间[i-d,i)上达到k个赞
            flag[p[i].id]=1;
    }
    for(int i=0;i<N;i++)
        if(flag[i]==1) printf("%d\n",i);
    return 0;
}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第4章日志统计.java
//日志统计 2018年第九届省赛C/C++大学A组  lanqiaoOJ 179
/*暴力法。
重点在于所有帖子的排序。下面的Java代码，36~38行的排序，先按所有帖子的id排序，然后对同一个id，对它的点赞时间排序。经过排序之后，所有帖子按id被分开，同一个id的帖子又按时间排序，容易统计是否为热帖。
在主程序中，暴力遍历所有排序后的帖子即可：第16~20行，对某个id的所有点赞贴，看是否存在某个[T, T+D)内达到了k个赞。
算法的时间复杂度，第16行的for循环是O(n)的。前面的排序是O(nlogn)的。总复杂度O(nlogn)。*/


import java.util.*;
public class Main{
    public static void main(String args[])    {
        int n,d,k;
        Scanner sc=new Scanner(System.in);
        n=sc.nextInt();
        d=sc.nextInt();
        k=sc.nextInt();
        post arr[]=new post[n];
        for(int i=0;i<n;i++)
            arr[i]=new post(sc.nextInt(),sc.nextInt());
        Arrays.sort(arr);
        int curid = arr[0].id;
        int added = 0;
        for(int i=0;i<n;i++){
            if(i+k-1<n && arr[i+k-1].id==curid 
                       && arr[i+k-1].ts-arr[i].ts<d && added==0){
                System.out.println(curid);
                added=1;
            }
            else if(arr[i].id != curid){
                curid = arr[i].id;
                added = 0;
                i = i-1;
            }
        }
    }
}
class post implements Comparable<post>{
    int ts,id;
    post(int a,int b)    {
        ts=a;
        id=b;
    }
    public int compareTo(post o) {
        if(id==o.id)   return ts-o.ts;
        else           return id-o.id;
    }
}

